README.rst
txgithub
========

txgithub is a library based on Twisted for accessing GitHub's API version 3.


Changelog
Changelog for txgitub
=====================

15.0.0 2015-01-12
----------------

* Allow specifying a context for commit statuses.
* Don't provide default description or URL for commit statuses.

0.2.0 2014-08-14
----------------

* Allow using a custom base URL to support private instances.


0.1.0 2013-03-13
----------------

* Allow repository status management.
* Allow repository hook management.
* Initial release.


txgithub/tests/test_api.py
"""
Tests for Twisted GitHub API bindings.
"""
from twisted.trial.unittest import SynchronousTestCase

from twisted.internet.defer import succeed

from txgithub.api import GithubApi as GitHubAPI


class TestReposEndpoint(SynchronousTestCase):
    """
    Unit tests for ReposEndpoint.
    """

    def _mock_makeRequest(self, *args, **kwargs):
        try:
            self._github_requests.append({'args': args, 'kwargs': kwargs})
            response = self._github_responses.pop()
            return succeed(response)
        except IndexError:
            return succeed({})

    def setUp(self):
        super(TestReposEndpoint, self).setUp()
        self._github_responses = []
        self._github_requests = []
        self.github = GitHubAPI(oauth2_token='fake-token')
        self.github.makeRequest = self._mock_makeRequest
        self.repos = self.github.repos

    def test_getHook_ok(self):
        """
        getHook return the info for a single hook.
        """
        self._github_responses = [{'hook': 'data'}]
        d = self.repos.getHook('repo', 'name', 123)

        result = self.successResultOf(d)

        self.assertEqual({'hook': 'data'}, result)
        request = self._github_requests[0]
        self.assertEqual('GET', request['kwargs']['method'])
        self.assertEqual(
            ['repos', 'repo', 'name', 'hooks', '123'],
            request['args'][0])

    def test_editHook_no_options(self):
        """
        editHook can be used for updating only the hook configuration,
        without updating events lists or active state.
        """
        self._github_responses = [{'updated': 'hook'}]
        config = {'url': 'some_url', 'content_type': 'json'}
        d = self.repos.editHook('repo', 'name', 123, 'web', config)

        result = self.successResultOf(d)

        self.assertEqual({'updated': 'hook'}, result)
        request = self._github_requests[0]
        self.assertEqual('PATCH', request['kwargs']['method'])
        self.assertEqual(
            ['repos', 'repo', 'name', 'hooks', '123'],
            request['args'][0])
        self.assertEqual(
            {
                'config': {'content_type': 'json', 'url': 'some_url'},
                'name': 'web',
            },
            request['kwargs']['post'])

    def test_editHook_with_options(self):
        """
        editHook can update the hooks state together with the list of
        active events.
        """
        config = {'url': 'some_url', 'content_type': 'json'}
        self.repos.editHook('repo', 'name', 123, 'web', config,
            active=False,
            events=['push'],
            add_events=['status'],
            remove_events=[],
            )

        request = self._github_requests[0]
        self.assertEqual(
            {
                'name': 'web',
                'active': False,
                'events': ['push'],
                'add_events': ['status'],
                'remove_events': [],
                'config': {'content_type': 'json', 'url': 'some_url'},
                'name': 'web',
            },
            request['kwargs']['post'])

    def test_testHook_ok(self):
        """
        testHook just triggers the test and returns nothing.
        """
        self._github_responses = ['']
        d = self.repos.testHook('repo', 'name', 123)

        result = self.successResultOf(d)

        self.assertEqual('', result)
        request = self._github_requests[0]
        self.assertEqual('POST', request['kwargs']['method'])
        self.assertEqual(
            ['repos', 'repo', 'name', 'hooks', '123', 'tests'],
            request['args'][0])

    def test_getStatuses_ok(self):
        """
        getStatuses will return the raw dictionary returned by GitHub.
        """
        self._github_responses = [{'response': 1}]
        d = self.repos.getStatuses('repo', 'name', 'sha123')

        result = self.successResultOf(d)

        self.assertEqual({'response': 1}, result)
        request = self._github_requests[0]
        self.assertEqual('GET', request['kwargs']['method'])

    def test_createStatus_default(self):
        """
        When no target_url, description, or context is provided, createStatus()
        creates a status without them.
        as description.
        """
        d = self.repos.createStatus(
            'repo', 'name', 'sha123', 'success')

        self.successResultOf(d)

        self.assertEqual(1, len(self._github_requests))
        request = self._github_requests[0]
        self.assertEqual('POST', request['kwargs']['method'])
        self.assertNotIn('description', request['kwargs']['post'])
        self.assertNotIn('target_url', request['kwargs']['post'])
        self.assertNotIn('context', request['kwargs']['post'])

    def test_createStatus_explicit(self):
        """
        When no target_url and description is provided, createStatus()
        creates a status linking to API url and having the state name
        as description.
        """
        d = self.repos.createStatus(
            'repo', 'name', 'sha123', 'success',
            description="desc", context="test-context",
            target_url="http://example.com/")

        self.successResultOf(d)

        self.assertEqual(1, len(self._github_requests))
        request = self._github_requests[0]
        self.assertEqual('POST', request['kwargs']['method'])
        self.assertEqual('desc', request['kwargs']['post']['description'])
        self.assertEqual('http://example.com/',
                         request['kwargs']['post']['target_url'])
        self.assertEqual('test-context', request['kwargs']['post']['context'])


txgithub/token.py
import os
import json
import base64

from twisted.web import client
from twisted.internet import utils

from txgithub.constants import HOSTED_BASE_URL

def createToken(username, password,
                note, note_url,
                scopes, baseURL=None):
    baseURL = baseURL or HOSTED_BASE_URL
    if baseURL[-1] != '/':
        baseURL += '/'

    raw = "%s:%s" % (username, password)
    encoded = base64.b64encode(raw).strip()
    headers = { 'Authorization' : 'Basic ' + encoded }

    postData = json.dumps(dict(
        note = note,
        note_url = note_url,
        scopes = scopes,
        ))
    d = client.getPage(
            url='%sauthorizations' % baseURL,
            method='POST',
            postdata=postData,
            headers=headers
            )
    @d.addCallback
    def extractToken(res):
        result = json.loads(res)
        return result['token']
    return d


def getToken():
    d = utils.getProcessOutput('git', ('config', '--get', 'github.token'), env=os.environ)
    d.addCallback(str.strip)
    return d


txgithub/constants.py
# This file is part of txgithub.  txgithub is free software: you can
# redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright Buildbot Team Members

"""
Common constants
"""

# Endpoint for hosted GitHub
HOSTED_BASE_URL = 'https://api.github.com/'


txgithub/scripts/gist.py
import sys
from os import path
from twisted.python import usage
from twisted.internet import defer

class Options(usage.Options):
    synopsis = "[-t <token>] <files>"
    optParameters = [["token", "t", None, "oauth token"]]

    longdesc = "Posts a gist."

    def parseArgs(self, *files):
        self['files'] = files

@defer.inlineCallbacks
def postGist(reactor, token, files):
    if not token:
        from txgithub import token
        token = yield token.getToken()

    from txgithub import api
    api = api.GithubApi(token)

    gistFiles = {}
    if files:
        for name in files:
            with open(name) as f:
                gistFiles[path.basename(name)] = {"content": f.read()}
    else:
        gistFiles['gistfile1'] = {"content": sys.stdin.read()}

    response = yield api.gists.create(files=gistFiles)
    print response['html_url']


def run(reactor, *argv):
    config = Options()
    try:
        config.parseOptions(argv[1:]) # When given no argument, parses sys.argv[1:]
    except usage.UsageError, errortext:
        print '%s: %s' % (argv[0], errortext)
        print '%s: Try --help for usage details.' % (argv[0])
        sys.exit(1)

    return postGist(reactor, **config)


txgithub/scripts/create_token.py
from __future__ import print_function

import sys
from twisted.python import usage

class Options(usage.Options):
    optParameters = [["note", "n", "txgithub", "token note"],
            ["url", "u", "https://github.com/tomprince/txgithub", "token note url"]
            ]

    longdesc = "Create a new github oauth2 token."

    def __init__(self):
        usage.Options.__init__(self)
        self['scopes'] = []

    def opt_scope(self, scope):
        self['scopes'].append(scope)
    opt_s = opt_scope

    def parseArgs(self, username):
        self['username'] = username

def createToken(reactor, username, password, note, url, scopes):
    from txgithub import token
    d = token.createToken(username, password,
            note=note, note_url=url,
            scopes = scopes
            )
    d.addCallback(print)
    return d

def run(reactor, *argv):
    config = Options()
    try:
        config.parseOptions(argv[1:]) # When given no argument, parses sys.argv[1:]
    except usage.UsageError, errortext:
        print('%s: %s' % (argv[0], errortext))
        print('%s: Try --help for usage details.' % (argv[0]))
        sys.exit(1)

    import getpass
    config['password'] = getpass.getpass("github password: ")

    return createToken(reactor, **config)


txgithub/api.py
# This file is part of txgithub.  txgithub is free software: you can
# redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright Buildbot Team Members

import re
import json
from twisted.python import log
from twisted.internet import defer, ssl
from twisted.web import client

from txgithub.constants import HOSTED_BASE_URL

class _GithubPageGetter(client.HTTPPageGetter):

    def handleStatus_204(self):
        # github returns 204 for e.g., DELETE operations
        self.handleStatus_200()

class _GithubHTTPClientFactory(client.HTTPClientFactory):

    protocol = _GithubPageGetter

    # dont' log about starting and stopping
    noisy = False

class GithubApi(object):
    # Interface to the github API, using
    # - API v3
    # - optional user/pass auth (token is not available with v3)
    # - async API

    def __init__(self, oauth2_token, baseURL=None):
        self._baseURL = baseURL or HOSTED_BASE_URL
        self.oauth2_token = oauth2_token
        self.rateLimitWarningIssued = False
        self.contextFactory = ssl.ClientContextFactory()

    def _makeHeaders(self):
        assert self.oauth2_token, "no token specified"
        return { 'Authorization' : 'token ' + self.oauth2_token }

    def makeRequest(self, url_args, post=None, method='GET', page=0):
        headers = self._makeHeaders()

        url = self._baseURL
        url += '/'.join(url_args)
        if page:
            url += "?page=%d" % page

        postdata = None
        if post:
            postdata = json.dumps(post)

        log.msg("fetching '%s'" % (url,), system='github')
        factory = _GithubHTTPClientFactory(url, headers=headers,
                    postdata=postdata, method=method,
                    agent='txgithub', followRedirect=0,
                    timeout=30)
        from twisted.internet import reactor
        reactor.connectSSL(factory.host, factory.port, factory,
                           self.contextFactory)
        d = factory.deferred
        @d.addCallback
        def check_ratelimit(data):
            self.last_response_headers = factory.response_headers
            remaining = int(factory.response_headers.get(
                                    'x-ratelimit-remaining', [0])[0])
            if remaining < 100 and not self.rateLimitWarningIssued:
                log.msg("warning: only %d Github API requests remaining "
                        "before rate-limiting" % remaining)
                self.rateLimitWarningIssued = True
            return data
        @d.addCallback
        def un_json(data):
            if data:
                return json.loads(data)
        return d

    link_re = re.compile('<([^>]*)>; rel="([^"]*)"')
    @defer.inlineCallbacks
    def makeRequestAllPages(self, url_args):
        page = 0
        data = []
        while True:
            data.extend((yield self.makeRequest(url_args, page=page)))
            if 'link' not in self.last_response_headers:
                break
            link_hdr = self.last_response_headers['link']
            for link in self.link_re.findall(link_hdr):
                if link[1] == 'next':
                    # note that we don't *use* the page -- why bother?
                    break
            else:
                break # no 'next' link, so we're done
            page += 1
        defer.returnValue(data)

    _repos = None
    @property
    def repos(self):
        if not self._repos:
            self._repos = ReposEndpoint(self)
        return self._repos

    _gists = None
    @property
    def gists(self):
        if not self._gists:
            self._gists = GistsEndpoint(self)
        return self._gists


class BaseEndpoint(object):

    def __init__(self, api):
        self.api = api


class ReposEndpoint(BaseEndpoint):

    @defer.inlineCallbacks
    def getEvents(self, repo_user, repo_name, until_id=None):
        """Get all repository events, following paging, until the end
        or until UNTIL_ID is seen.  Returns a Deferred."""
        done = False
        page = 0
        events = []
        while not done:
            new_events = yield self.api.makeRequest(
                    ['repos', repo_user, repo_name, 'events'],
                    page)

            # terminate if we find a matching ID
            if new_events:
                for event in new_events:
                    if event['id'] == until_id:
                        done = True
                        break
                    events.append(event)
            else:
                done = True

            page += 1
        defer.returnValue(events)

    def getHooks(self, repo_user, repo_name):
        """Get all repository hooks.  Returns a Deferred."""
        return self.api.makeRequestAllPages(
            ['repos', repo_user, repo_name, 'hooks'])

    def getHook(self, repo_user, repo_name, hook_id):
        """
        GET /repos/:owner/:repo/hooks/:id

        Returns the Hook.
        """
        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'hooks', str(hook_id)],
            method='GET',
            )

    def createHook(self, repo_user, repo_name, name, config, events, active):
        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'hooks'],
            method='POST',
            post=dict(name=name, config=config, events=events, active=active))

    def editHook(self, repo_user, repo_name, hook_id, name, config,
            events=None, add_events=None, remove_events=None, active=None):
        """
        PATCH /repos/:owner/:repo/hooks/:id

        :param hook_id: Id of the hook.
        :param name:  The name of the service that is being called.
        :param config: A Hash containing key/value pairs to provide settings
                       for this hook.
        """
        post = dict(
                name=name,
                config=config,
                )
        if events is not None:
            post['events'] = events

        if add_events is not None:
            post['add_events'] = add_events

        if remove_events is not None:
            post['remove_events'] = remove_events

        if active is not None:
            post['active'] = active

        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'hooks', str(hook_id)],
            method='PATCH',
            post=post,
            )

    def testHook(self, repo_user, repo_name, hook_id):
        """
        POST /repos/:owner/:repo/hooks/:id/tests

        Response headers:
            Status: 204 No Content
            X-RateLimit-Limit: 5000
            X-RateLimit-Remaining: 4999
        Response content:
            None
        """
        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'hooks', str(hook_id), 'tests'],
            method='POST',
            )

    def deleteHook(self, repo_user, repo_name, id):
        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'hooks', str(id)],
            method='DELETE')

    def getStatuses(self, repo_user, repo_name, sha):
        """
        :param sha: Full sha to list the statuses from.
        :return: A defered with the result from GitHub.
        """
        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'statuses', sha],
            method='GET')

    def createStatus(self,
            repo_user, repo_name, sha, state, target_url=None,
            description=None, context=None):
        """
        :param sha: Full sha to create the status for.
        :param state: one of the following 'pending', 'success', 'error'
                      or 'failure'.
        :param target_url: Target url to associate with this status.
        :param description: Short description of the status.
        :return: A defered with the result from GitHub.
        """
        payload = {'state': state}

        if description is not None:
            payload['description'] = description

        if target_url is not None:
            payload['target_url'] = target_url

        if context is not None:
            payload['context'] = context

        return self.api.makeRequest(
            ['repos', repo_user, repo_name, 'statuses', sha],
            method='POST',
            post=payload)


class GistsEndpoint(BaseEndpoint):
    def create(self, files, description=None, public=True):
        data = { 'files': files, 'public': bool(public) }
        if description is not None:
            data['description'] = description
        return self.api.makeRequest(
                ['gists'],
                method='POST',
                post=data)
